pytest_lazyfixture.py

```diff
@@ -49,10 +49,21 @@ def pytest_fixture_setup(fixturedef, request):
 
 
 def pytest_runtest_call(item):
+    def _rec_part(val):
+        if is_lazy_fixture(val):
+            return item._request.getfixturevalue(val.name)
+        elif type(val) == tuple:
+            return tuple(item._request.getfixturevalue(v.name) if is_lazy_fixture(v) else _rec_part(v) for v in val)
+        elif type(val) == list:
+            return list(item._request.getfixturevalue(v.name) if is_lazy_fixture(v) else _rec_part(v) for v in val)
+        elif isinstance(val, dict):
+            return {key: item._request.getfixturevalue(v.name) if is_lazy_fixture(v) else _rec_part(v)
+                    for key, v in val.items()}
+        return val
```

Comment:
Sorry is a problem with dictionaries only? Or with list too?
Actually it checks it now, so only some more levels are added :)

But I think decrease amount of checks is a good idea. But how to do it?
Add new attr to pytest.fixture?

     @pytest.fixture(params=[. . .], check_collections_for_lazy_fixtures=True)

or add specific object: 

    @pytest.fixture(params=[check_collections_for_lazy_fixtures, . . .])

Will it works?

Reply:
I found the better option:

add an option to pytest.ini lookup for lazy_fixtures (turn off by default).
using decorator @pytest.mark.lookup_for_lazy_fixtures().

If global config is turned on the we always look, if it turned off we check for decorator in item.own_markers.

BTW do you concern about slowing down performance, because of lazy_fixtures lookup?
I am not sure it will have a big impact, because usually (at least on project I worked on :) ) test doesn't use big collections with a lot of  nested collections inside.

Reply:
Yeah, I think about using marks too. I will take a look a little bit later though.

> BTW do you concern about slowing down performance, because of lazy_fixtures lookup?

Yep :) Also I don't like unknown structure of the param. Maybe it will be a good idea to specify format so we will know where to lookup lazy_fixture?

Reply:
I am trying to run some benchmark, but something broke:

     C:\Users\artem\PycharmProjects\pytest-lazy-fixture\tests\test_lazyfixture.py:18: 
     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
     C:\Program Files\python\lib\site-packages\_pytest\pytester.py:993: in getitems
          modcol = self.getmodulecol(source)
     C:\Program Files\python\lib\site-packages\_pytest\pytester.py:1020: in getmodulecol
        self.config = config = self.parseconfigure(path, *configargs)
     C:\Program Files\python\lib\site-packages\_pytest\pytester.py:959: in parseconfigure
        config._do_configure()
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    self = <_pytest.config.Config object at 0x0000016F434B08C8>

    def _do_configure(self):
     >       assert not self._configured
     E       AssertionError

     C:\Program Files\python\lib\site-packages\_pytest\config\__init__.py:634: AssertionError

Do you have any idea what's wrong?

>Also I don't like unknown structure of the param. Maybe it will be a good idea to specify format so we will know where to lookup lazy_fixture?

I got your concern, but I think specifying format is a source of many possible format, the users can muddle up or forget to specify it, also the can be a problem with parsing the format, and it can create bugs in lib, also that can be slower than just checking all collections. BTW we decided to keep it simple and specifying format is not very intuitive and convenient for users.